test/unit/math/opencl/copy_test.cpp

```diff
@@ -8,6 +8,151 @@
 #include <algorithm>
 #include <vector>
 
+TEST(MathMatrixGPU, copy_destroyed_constructor_eval) {
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::MatrixXd;
+  using stan::math::from_matrix_cl;
+  using stan::math::matrix_cl;
+  int N = 100;
+  MatrixXd a = MatrixXd::Random(N, N);
+  MatrixXd b = MatrixXd::Random(N, N);
+  matrix_cl<double> c_cl(
+      (a + b).eval());  // the problem, this case with .eval()
+  // attempt to scramble the memory that was used by the temporary
+  MatrixXd w = a - b;
+
+  // make compiler think a and b were changed, so second sum can not be
+  // optimized away
+  volatile int i = 19;
+  volatile double no_opt = a(i);
+  a(i) = no_opt;
+  no_opt = b(i);
+  b(i) = no_opt;
+
+  MatrixXd correct = a + b;
+  MatrixXd result = from_matrix_cl(c_cl);
+  for (int i = 0; i < N * N; i++) {
+    EXPECT_EQ(result(i), correct(i));
+  }
+  no_opt = w.array().sum();
+}
+
+TEST(MathMatrixGPU, copy_destroyed_to_matrix_cl_eval) {
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::MatrixXd;
+  using stan::math::from_matrix_cl;
+  using stan::math::matrix_cl;
+  int N = 100;
+  MatrixXd a = MatrixXd::Random(N, N);
+  MatrixXd b = MatrixXd::Random(N, N);
+  matrix_cl<double> c_cl
+      = stan::math::to_matrix_cl((a + b).eval());  // the problem
+  // attempt to scramble the memory that was used by the temporary
+  MatrixXd w = a - b;
+
+  // make compiler think a and b were changed, so second sum can not be
+  // optimized away
+  volatile int i = 19;
+  volatile double no_opt = a(i);
+  a(i) = no_opt;
+  no_opt = b(i);
+  b(i) = no_opt;
+
+  MatrixXd correct = a + b;
+  MatrixXd result = from_matrix_cl(c_cl);
+  for (int i = 0; i < N * N; i++) {
+    EXPECT_EQ(result(i), correct(i));
+  }
+  no_opt = w.array().sum();
+}
+
+TEST(MathMatrixGPU, copy_destroyed_to_matrix_cl_scalar) {
+  using Eigen::Dynamic;
+  using Eigen::Matrix;
+  using Eigen::MatrixXd;
+  using stan::math::from_matrix_cl;
+  using stan::math::matrix_cl;
+  double a = 1.0;
+  double b = 2.0;
+  double c = 3.0;
+  double d = 4.0;
+  matrix_cl<double> c_cl
+      = stan::math::to_matrix_cl(a + b + c + d);  // the problem
+  // attempt to scramble the memory that was used by the temporary
+  double f = a - b - c - d;
```

Comment:
Did this test reproduce the bug for you? I think compiler is likely to only use registers for scalars, so we have little or no chance to actually overwrite them.

Reply:
Yes, this reproduces the bug on my notebook. I didnt try on other machines because I didnt have access to any other today. It behaves like its undefined, 1E-316 and such.